### PR TITLE
Add missing fields to generated Deploy methods

### PR DESCRIPTION
**Pull request by jarnaud**
## Motivation

In `gethwrappers`, the generated `Deploy<contract>` methods do not fill the address and abi fields.

Eg:
```
func DeployEVM2EVMOffRampHelper(auth *bind.TransactOpts, backend bind.ContractBackend, staticConfig EVM2EVMOffRampStaticConfig, sourceTokens []common.Address, pools []common.Address, rateLimiterConfig RateLimiterConfig) (common.Address, *types.Transaction, *EVM2EVMOffRampHelper, error) {
	parsed, err := EVM2EVMOffRampHelperMetaData.GetAbi()
	if err != nil {
		return common.Address{}, nil, nil, err
	}
	if parsed == nil {
		return common.Address{}, nil, nil, errors.New("GetABI returned nil")
	}

	address, tx, contract, err := bind.DeployContract(auth, *parsed, common.FromHex(EVM2EVMOffRampHelperBin), backend, staticConfig, sourceTokens, pools, rateLimiterConfig)
	if err != nil {
		return common.Address{}, nil, nil, err
	}
	return address, tx, &EVM2EVMOffRampHelper{EVM2EVMOffRampHelperCaller: EVM2EVMOffRampHelperCaller{contract: contract}, EVM2EVMOffRampHelperTransactor: EVM2EVMOffRampHelperTransactor{contract: contract}, EVM2EVMOffRampHelperFilterer: EVM2EVMOffRampHelperFilterer{contract: contract}}, nil
}
```

We need to reach this state:

```
func DeployEVM2EVMOffRampHelper(auth *bind.TransactOpts, backend bind.ContractBackend, staticConfig EVM2EVMOffRampStaticConfig, sourceTokens []common.Address, pools []common.Address, rateLimiterConfig RateLimiterConfig) (common.Address, *types.Transaction, *EVM2EVMOffRampHelper, error) {
	parsed, err := EVM2EVMOffRampHelperMetaData.GetAbi()
	if err != nil {
		return common.Address{}, nil, nil, err
	}
	if parsed == nil {
		return common.Address{}, nil, nil, errors.New("GetABI returned nil")
	}

	address, tx, contract, err := bind.DeployContract(auth, *parsed, common.FromHex(EVM2EVMOffRampHelperBin), backend, staticConfig, sourceTokens, pools, rateLimiterConfig)
	if err != nil {
		return common.Address{}, nil, nil, err
	}
	return address, tx, &EVM2EVMOffRampHelper{address: address, abi: *parsed, EVM2EVMOffRampHelperCaller: EVM2EVMOffRampHelperCaller{contract: contract}, EVM2EVMOffRampHelperTransactor: EVM2EVMOffRampHelperTransactor{contract: contract}, EVM2EVMOffRampHelperFilterer: EVM2EVMOffRampHelperFilterer{contract: contract}}, nil
}
```

## Solution

Modify the generator to add the address and abi fields in the returned struct.

### CODE REVIEW

File: core/gethwrappers/abigen_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+// Test running an abigen command to generate a contract wrapper.
+// Using as test example:
+// go run ./generation/generate/wrap.go ../../contracts/solc/v0.6/VRF.abi ../../contracts/solc/v0.6/VRF.bin VRF solidity_vrf_wrapper
```
Comment (RensR) [inline core/gethwrappers/abigen_test.go:12]: 0.6 will be removed from this repo soon, could we use a 0.8 example?

File: core/gethwrappers/abigen_test.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+func TestAbigen(t *testing.T) {
+	abiPath := "../../contracts/solc/v0.6/VRF.abi"
+	binPath := "../../contracts/solc/v0.6/VRF.bin"
+	className := "VRF"
+	pkgName := "solidity_vrf_wrapper"
+	fmt.Println("Generating", pkgName, "contract wrapper")
+
+	cwd, err := os.Getwd() // gethwrappers directory
+	if err != nil {
```
Comment (RensR) [inline core/gethwrappers/abigen_test.go:21]: require.NoError(t, err)

is the normal way of checking errors in tests

File: core/gethwrappers/abigen_test.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		Exit("could not get working directory", err)
+	}
+	outDir := filepath.Join(cwd, "generated", pkgName)
+	if mkdErr := os.MkdirAll(outDir, 0700); err != nil {
+		Exit("failed to create wrapper dir", mkdErr)
+	}
+	outPath := filepath.Join(outDir, pkgName+".go")
+
+	Abigen(AbigenArgs{
+		Bin: binPath, ABI: abiPath, Out: outPath, Type: className, Pkg: pkgName,
```
Comment (RensR) [inline core/gethwrappers/abigen_test.go:31]: I would expect a test to verify that the newly added fields are not empty

File: core/gethwrappers/abigen_test.go (hunk 1 above)
Comment (jarnaud) [reply core/gethwrappers/abigen_test.go]: Now using the LogEmitter instead.

File: core/gethwrappers/abigen_test.go (hunk 3 above)
Comment (jarnaud) [reply core/gethwrappers/abigen_test.go]: Changed the test to call the `Address()` method to check address correctness.

**Review by mateusz-sekara — APPROVED** [review]
Looks good to me, but it would be great for someone from the Foundations team to take a look as well. @jmank88 maybe?

File: core/gethwrappers/abigen.go [hunk 4]
```diff
@@ -264,6 +264,41 @@ func addContractStructFields(contractName string, fileNode *ast.File) *ast.File
...
+
+		for _, stmt := range x.Body.List {
+			returnStmt, is := stmt.(*ast.ReturnStmt)
+			if !is {
+				continue
+			}
+			if len(returnStmt.Results) < 3 {
+				continue
+			}
+			lit, is := returnStmt.Results[2].(*ast.UnaryExpr).X.(*ast.CompositeLit)
```
Comment (jmank88) [inline core/gethwrappers/abigen.go:284]: Could the first cast to `*ast.UnaryExpr` fail?

**Review by jmank88 — APPROVED** [review]

File: core/gethwrappers/abigen.go (hunk 4 above)
Comment (jarnaud) [reply core/gethwrappers/abigen.go]: I added a check to this type assertion.
Also refactored the code a bit.

**Review by mateusz-sekara — APPROVED** [review]

File: core/gethwrappers/abigen.go [hunk 5]
```diff
@@ -260,11 +268,49 @@ func addContractStructFields(contractName string, fileNode *ast.File) *ast.File
...
+				addressExpr := &ast.KeyValueExpr{
+					Key:   ast.NewIdent("address"),
+					Value: ast.NewIdent("address"),
+				}
+				abiExpr := &ast.KeyValueExpr{
+					Key:   ast.NewIdent("abi"),
+					Value: ast.NewIdent("*parsed"),
+				}
+				lit.Elts = append([]ast.Expr{addressExpr, abiExpr}, lit.Elts...)
+			}
```
[the remark this answers is not in the document]
Comment (jarnaud) [reply core/gethwrappers/abigen.go]: Indeed, removed the useless `else`

**Review by mateusz-sekara — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Pull request merged**